### PR TITLE
Add GATK's CreateReadCountPanelOfNormals and DenoiseReadCounts

### PR DESCRIPTION
**Pull request by ramprasadn**
## PR checklist

See title.

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [x] Emit the `versions.yml` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`
  - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware`

### CODE REVIEW

File: modules/nf-core/gatk4/denoisereadcounts/main.nf [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        'biocontainers/gatk4:4.4.0.0--py36hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(counts)
+    tuple val(meta2), path(pon)
+
+    output:
+    tuple val(meta), path("*_standardizedCR.tsv"), emit: standardized
+    tuple val(meta), path("*_denoisedCR.tsv"),     emit: denoised
+    path "versions.yml"           , emit: versions
```
Comment (sguizard) [inline modules/nf-core/gatk4/denoisereadcounts/main.nf:17]:
```suggestion
    tuple val(meta_count), path("*_standardizedCR.tsv"), emit: standardized
    tuple val(meta_count), path("*_denoisedCR.tsv")    , emit: denoised
    path "versions.yml"                          , emit: versions
```

File: modules/nf-core/gatk4/denoisereadcounts/main.nf [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    label 'process_single'
+
+    conda "bioconda::gatk4=4.4.0.0"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/gatk4:4.4.0.0--py36hdfd78af_0':
+        'biocontainers/gatk4:4.4.0.0--py36hdfd78af_0' }"
+
+    input:
+    tuple val(meta), path(counts)
+    tuple val(meta2), path(pon)
```
Comment (sguizard) [inline modules/nf-core/gatk4/denoisereadcounts/main.nf:12]:
```suggestion
    tuple val(meta_count), path(counts)
    tuple val(meta_pon)  , path(pon)
```
Less confusing than meta/meta2

File: modules/nf-core/gatk4/denoisereadcounts/main.nf [hunk 3]
```diff
@@ -0,0 +1,57 @@
+process GATK4_DENOISEREADCOUNTS {
+    tag "$meta.id"
```
Comment (sguizard) [inline modules/nf-core/gatk4/denoisereadcounts/main.nf:2]:
```suggestion
    tag "$meta_count.id"
```

File: modules/nf-core/gatk4/denoisereadcounts/main.nf [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    tuple val(meta), path("*_standardizedCR.tsv"), emit: standardized
+    tuple val(meta), path("*_denoisedCR.tsv"),     emit: denoised
+    path "versions.yml"           , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (sguizard) [inline modules/nf-core/gatk4/denoisereadcounts/main.nf:24]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta_count.id}"
```

File: modules/nf-core/gatk4/denoisereadcounts/main.nf [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        gatk4: \$(echo \$(gatk --version 2>&1) | sed 's/^.*(GATK) v//; s/ .*\$//')
+    END_VERSIONS
+    """
+
+    stub:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
```
Comment (sguizard) [inline modules/nf-core/gatk4/denoisereadcounts/main.nf:47]:
```suggestion
    def prefix = task.ext.prefix ?: "${meta_count.id}"
```

File: modules/nf-core/gatk4/denoisereadcounts/meta.yml [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  - meta2:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. `[ id:'test', single_end:false ]`
+  - counts:
+      type: file
+      description: Read counts in hdf5 or tsv format.
+      pattern: "*.{hdf5,tsv}"
+
```
Comment (sguizard) [inline modules/nf-core/gatk4/denoisereadcounts/meta.yml:34]:
```suggestion
 - pon:
      type: file
      description: Panel-Of-Normals file.
      pattern: "*.hdf5"
```
Missing pon input

File: modules/nf-core/gatk4/denoisereadcounts/meta.yml [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+output:
+  - meta:
```
Comment (sguizard) [inline modules/nf-core/gatk4/denoisereadcounts/meta.yml:36]:
```suggestion
  - meta_count:
```

**Review by sguizard — CHANGES_REQUESTED** [review]
One suggestion of code update about `meta`/`meta2` in `denoisereadcounts/main.nf`. 
I'm not sure if there is any official rules about naming additional meta input. I leave you the choice to apply them or not. 
One mandatory update. The `pon` input is missing from `denoisereadcounts/meta.yml`. 
Otherwise, LGTM. :+1:

**Comment by ramprasadn** [comment]
Thanks for the review @sguizard. In terms of naming meta, I think that's a good suggestion. I am leaving things as it is, because from what I have seen naming meta as `meta, meta2, meta3....` appears to be the standard (though its not written down in the guidelines). But perhaps would be nice to initiate a conversation among other devs, and see what they think 😄

**Comment by ramprasadn** [comment]
All the tests are passing now. Not sure what's going on with the module denoisereadcounts randomly failing CI, and then successfully finishing upon restarting the failed test. But at least its clear that the issue is not with the code itself. 😄

File: modules/nf-core/gatk4/createreadcountpanelofnormals/main.nf [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args       = task.ext.args ?: ''
+    def prefix     = task.ext.prefix ?: "${meta.id}"
+    def input_list = counts.collect(){"--input $it"}.join(" ")
+
+    def avail_mem = 3072
+    if (!task.memory) {
+        log.info '[GATK CondenseDepthEvidence] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
```
Comment (nvnieuwk) [inline modules/nf-core/gatk4/createreadcountpanelofnormals/main.nf:27]:
```suggestion
        log.info '[GATK CreateReadCountPanelOfNormals] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
```

File: modules/nf-core/gatk4/createreadcountpanelofnormals/main.nf [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    } else {
+        avail_mem = (task.memory.mega*0.8).intValue()
+    }
+    """
+    gatk --java-options "-Xmx${avail_mem}M" CreateReadCountPanelOfNormals \\
+        ${input_list} \\
```
Comment (nvnieuwk) [inline modules/nf-core/gatk4/createreadcountpanelofnormals/main.nf:33]:
```suggestion
        ${args} \\
        ${input_list} \\
```
Don't forget the args ;)

File: modules/nf-core/gatk4/denoisereadcounts/main.nf (hunk 2 above)
Comment (nvnieuwk) [reply modules/nf-core/gatk4/denoisereadcounts/main.nf:12]: This is done like this to keep all meta names uniform across all modules

File: modules/nf-core/gatk4/denoisereadcounts/main.nf [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+    def avail_mem = 3072
+    if (!task.memory) {
+        log.info '[GATK CondenseDepthEvidence] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
+    } else {
+        avail_mem = (task.memory.mega*0.8).intValue()
+    }
+    """
+    gatk --java-options "-Xmx${avail_mem}M" DenoiseReadCounts \\
+        --input ${counts} \\
```
Comment (nvnieuwk) [inline modules/nf-core/gatk4/denoisereadcounts/main.nf:34]:
```suggestion
        ${args} \\
        --tmp-dir . \\
        --input ${counts} \\
```

args again and added the temp dir to the current directory (this to make sure a limited temp folder does not cause the pipeline to fail)

File: modules/nf-core/gatk4/denoisereadcounts/main.nf [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+
+    def avail_mem = 3072
+    if (!task.memory) {
+        log.info '[GATK CondenseDepthEvidence] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
```
Comment (nvnieuwk) [inline modules/nf-core/gatk4/denoisereadcounts/main.nf:28]:
```suggestion
        log.info '[GATK DenoiseReadCounts] Available memory not known - defaulting to 3GB. Specify process memory requirements to change this.'
```

**Review by nvnieuwk — CHANGES_REQUESTED** [review]

**Comment by nvnieuwk** [comment]
> Thanks for the review @sguizard. In terms of naming meta, I think that's a good suggestion. I am leaving things as it is, because from what I have seen naming meta as `meta, meta2, meta3....` appears to be the standard (though its not written down in the guidelines). But perhaps would be nice to initiate a conversation among other devs, and see what they think smile

I don't think this will be an improvement. IMO it's better to keep the variables names of the meta uniform across all modules. (especially because you would need to look at the module code every time you try to use the meta)

File: modules/nf-core/gatk4/createreadcountpanelofnormals/main.nf (hunk 9 above)
Comment (ramprasadn) [reply modules/nf-core/gatk4/createreadcountpanelofnormals/main.nf:34]: Good catch! I copied this bit from another script, and I didn't realize it was missing `args` 😄

**Review by nvnieuwk — APPROVED** [review]
Awesome!

**Review by sguizard — APPROVED** [review]

**Comment by ramprasadn** [comment]
Thanks @sguizard @nvnieuwk  🚀

**Pull request merged**